### ISSUE TITLE
possible memory leak

### ISSUE THREAD

**Issue by vlukashov** [vaadin/vaadin-router#376]
The issue is reported on the Vaadin Forum: https://vaadin.com/tutorials/lit-element/navigation-and-code-splitting

> Changing the view between stats-view and todo-view creates a memory leak. The more times we change the view, the more memory the application uses.
>
>intensive change of views causes accumulating garbage:
>![image.png](https://images.zenhubusercontent.com/5a572ab64b5806bc2bc2536e/99abb017-24dd-4eda-a127-1c4b44ad032c)
>`dev tools > memory > allocation` instrumentation on timeline
>
>same intensive change of views but without dev tools (only task manager opened) Chrome 76 after several dozen view changes apps memory increase to 205MB. It doesn't look good
>
>i've used:
>
>```
>"dependencies": {
>  "lit-element": "2.2.1",
>  "@vaadin/router": "1.3.0",
>  "@webcomponents/webcomponentsjs": "2.2.10"
>}
>```
>
>Any idea where find the source of leak ?
>Is this a problem of `@vaadin/router` or `lit-element` ?

**Comment by Legioth** [comment]
Potentially related to https://github.com/vaadin/vaadin-charts/issues/445

**Comment by platosha** [comment]
Caused by the following factors:

1. On every navigation, Vaadin Router creates a new instance of components. This creates some garbage collector pressure.

2. Certain components are quite heavy, notably Vaadin Charts are causing the most of the memory consumption in the tutorial app. Removing Vaadin Charts import slows down the memory usage growth quite a bit.

While the latter is a separate issue, we could take some measures in Vaadin Router to enable reusing component instances.

Starting from v1.4.0, the Router supports routes with action functions that return the same component instance:

```javascript
// create your view only once
const myViewElement = document.createElement('my-view');

// use the same instance for every navigation
setRoutes([{path: '/my-view', action: () => myViewElement}]);
```

Unfortunately, this should be supported in the view component lifecycle, so that the view component updates and renders appropriately on navigation.

I’ve tried to upgrade `vaadin-router` to v1.4.0 and tweak the router configuration that way in the tutorial app. The memory increase stops completely, but the views appear broken or empty on second / third / subsequent navigations. Looks like there are issues with reusing LitElement-based component instances at the moment.

**Comment by platosha** [comment]
Tried to reproduce again, this time was clicking the “Collect garbage” button in Chrome DevTools. 

Without Vaadin Charts, after loading page, filling some data and collecting garbage, the original VM size was ~16 MB for me. During navigation it was briefly growing, however, after garbage collector, it shrinks back to the initial value.

With Vaadin Charts, the original VM size was ~20 MB, then after navigations and garbage collection, it didn’t reach to the original value. The garbage-collected VM size was increasing the more navigations I did, in the end was ~30 MB after a few dozen navigations.

This suggests that the issue is likely in Vaadin Charts.
